packages/react/toggle-button/src/ToggleButton.tsx

```diff
@@ -16,53 +205,347 @@ type ToggleButtonOwnProps = {
+      active: state.value === 'on',
     });
-
+    const allProps = mergeProps<typeof props>(buttonProps, rovingFocusProps);
     return (
-      <Comp
-        {...getPartDataAttrObj(NAME)}
-        type="button"
-        aria-pressed={toggled}
-        data-state={toggled ? 'on' : 'off'}
-        data-disabled={props.disabled ? '' : undefined}
-        ref={forwardedRef}
-        onClick={composeEventHandlers(onClick, () => {
-          if (!props.disabled) {
-            setToggled(!toggled);
-          }
-        })}
-        {...buttonProps}
-      >
+      <Comp {...getPartDataAttrObj(BUTTON_NAME)} ref={forwardedRef} {...allProps}>
         {children}
       </Comp>
     );
   }
 );
 
-ToggleButton.displayName = NAME;
+const ToggleButton = forwardRefWithAs<typeof BUTTON_DEFAULT_TAG, ToggleButtonOwnProps>(
+  (props, forwardedRef) => {
+    const context = useInternalToggleButtonGroupContext({
+      toggledProp: props.toggled,
+      defaultToggled: props.defaultToggled,
+    });
+
+    return context?.rovingFocus ? (
+      <RovingToggleButton ref={forwardedRef} {...props} />
+    ) : (
+      <ToggleButtonImpl ref={forwardedRef} {...props} />
+    );
+  }
+);
+
+ToggleButton.displayName = BUTTON_NAME;
+ToggleButtonGroup.displayName = GROUP_NAME;
+ToggleButtonGroupExclusive.displayName = GROUP_EXC_NAME;
 
 const Root = ToggleButton;
+const Group = ToggleButtonGroup;
+const GroupExclusive = ToggleButtonGroupExclusive;
 
-export {
-  ToggleButton,
-  //
-  Root,
-};
+export { ToggleButton, ToggleButtonGroup, ToggleButtonGroupExclusive, Root, Group, GroupExclusive };
+
+/* ---------------------------------------------------------------------------------------------- */
+
+function useToggleButton<
+  ElementType extends keyof JSX.IntrinsicElements | React.ComponentType<any>
+>(props: ToggleButtonOwnProps & React.ComponentProps<ElementType>) {
+  const {
+    toggled: toggledProp,
+    defaultToggled,
+    onClick,
+    onToggledChange,
+    children,
+    value: valueProp,
+    ...otherButtonProps
+  } = props;
+
+  const [_toggled = false, _setToggled] = useControlledState<boolean>({
+    prop: toggledProp,
+    onChange: onToggledChange,
+    defaultProp: defaultToggled || false,
+  });
+
+  const generatedValue = makeId(`toggle-button`, useId());
+  const value = valueProp || generatedValue;
+  const context = useInternalToggleButtonGroupContext({ toggledProp, defaultToggled });
+  const toggled = useInternalToggleButtonState({ context, ownToggledState: _toggled, value });
+
+  function setToggled(state: boolean) {
+    if (context) {
+      context.handleChange(value);
+    } else {
+      _setToggled(state);
+    }
+  }
+
+  // If the toggle button is in a group, onToggleChange will never fire because we bypass the
+  // setter returned from useControlledState since the group is managing its children states. We
+  // will explicitly call onToggleChange for grouped toggle buttons only when the toggled value
+  // changes.
```

Comment:
Comments like these are a red flag to me and further highlighting that we haven't separated our concerns appropriately. There are a bunch of hooks and hoops in this file that are getting really complex to follow for something seemingly simple 😬 

My preference is composition so that functionality is layered up as and when it is needed. Trying to squish everything in together is what causes theses branches/edge-cases that become hard to maintain and reason about.

Instead, I would envisage something like:

```jsx
// `ToggleGroupItem` doesn't have an `onToggleChange` callback because values 
// are managed by the group. This removes the `if else` context checking everywhere
// because the fact we're in a `ToggleGroupItem` gives us the intent implicitly.
function ToggleGroupItem() {
  return <ToggleButton onToggleChange={context.handleChange} />
}
```